design/design-doc-agent.md

```diff
@@ -0,0 +1,627 @@
+the secret whose name they specified on the `*Requirement` custom resource. To
+keep this experience, we need to have a synchronization loop for the following
+resources:
+
+* Pull
+  * `CustomResourceDefinition`s of all types that we want the applications to be
+    able to manage and view:
+    * Requirements that are published via `InfrastructurePublication`s. The
+      source of truth will be the remote cluster.
+    * `InfrastructureDefinition`, `InfrastructurePublication` and `Composition`.
+  * `Composition`s to discover what's available to choose. CRs of this type will
+    be read-only, and the source of truth will be the central cluster.
+  * `InfrastructurePublication`s to discover what's availabe as published API.
+    Read-only.
+  * `InfrastructureDefinition`s to discover how the secret keys are shaped.
+    Read-only.
+  * `Secret`s that are result of the infrastructure that is provisioned so that
+    it can be mounted to `Pod`s. Read-only.
+* Push
+  * `*Requirement` custom resources so that infrastructure can be requested.
+    Read and write permissions in a specific namespace in the central cluster
+    will be needed.
+
+### RBAC
+
+As we have two different Kubernetes clusters, there will be two separate security
+domains and because of that, the `ServiceAccount`s in the remote cluster will
+not be able to do any operation in the central cluster. Since the entity that
+will execute the operations in the central cluster is the agent, we need to
+define how we can deliver the necessary credentials to the agent so that it can
+connect to the central cluster. Additionally, it will need some permissions to
+execute operations in the remote cluster like `Secret` and
+`CustomResourceDefinition` creation. We will look at how the agent will be
+granted permissions to do its job in two separate domains with different
+mechanisms.
+
+#### Authenticating to The Central Cluster
+
+In order to execute any operation, a `ServiceAccount` needs to exist in the
+central cluster with appropriate permissions to read during pull and write
+during push operations while synchronizing with central cluster. Since the agent
+is running in the remote cluster, the credentials of this `ServiceAccount` will
+be stored in a `Secret` in the remote cluster. Alongside the credentials, the
+agent needs to know the namespace that it should sync to in the central cluster.
+
+The easiest configuration would be the one where we specify the `Secret` and a
+target namespace via installation commands of the agent. However, the namespace
+pairing list could change dynamically after the installation. So, we need to be
+able to supply the pairings after the installation. In order to specify which
+namespace in the remote cluster should be synced to which namespace in the
+central cluster, `Namespace` resource needs to be annotated.
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: foo
+  annotations:
+    "agent.crossplane.io/target-namespace": bar
```

Comment:
Would this annotation be mandatory, or optional?

Per our out of band conversation I suggest:

* The annotation be optional, and the agent default to trying to sync to the same namespace in the central cluster if the annotation is omitted. So if a requirement was authored in remote namespace `example` the agent would assume that an `example` namespace also existed in the central cluster and would try to create a corresponding requirement there.
* We avoid conflicts by using `metadata.generateName` when we create a corresponding requirement in the central cluster. That is, if the remote requirement is `{namespace: example, name: coolreq}` (and there were no annotations on the namespace) the corresponding central requirement would be `{namespace: example, generateName: coolreq}`.

We'd need to ensure there was some way for the remote requirement to regain control of the corresponding central requirement under this approach; presumably annotations creating a relationship at each end.

Reply:
> The annotation be optional, and the agent default to trying to sync to the same namespace in the central cluster if the annotation is omitted. So if a requirement was authored in remote namespace example the agent would assume that an example namespace also existed in the central cluster and would try to create a corresponding requirement there.

I have thought having that namespace pairing. I have a feeling that doing exact pairing of the namespaces could be invasive for one remote cluster's effect on that common central cluster. Let's say we have 100 remote clusters and they all have their own namespaces. Besides from conflicts of the general names that we all use when we create a namespace, I think an individual who is creating a namespace in one of those remote cluster may not know what he's signing up to that is claiming yet another full namespace in the central cluster. It just seems very in-directional information that they need to be aware of. So, I thought that if we let the admin of the remote cluster choose one namespace to be the default, then that individual will have less things to think/be careful about. That obviously increases the chance of conflict of the requirements but I'd say that's less of a concern compared to namespace conflicts which seems unrelated to a person setting up their own namespace. Also, changing the name of the requirement is much easier than changing the name of the namespace. In all cases, they can annotate the namespace in the remote cluster to avoid conflicts altogether.

> We avoid conflicts by using metadata.generateName when we create a corresponding requirement in the central cluster. That is, if the remote requirement is {namespace: example, name: coolreq} (and there were no annotations on the namespace) the corresponding central requirement would be {namespace: example, generateName: coolreq}.

I still couldn't convince myself for this; rejecting to sync and writing about it in the status sounds like a simpler approach without the complexity introduced by that name difference that will be caused by `generateName` and its second order effects like making the name even longer (original name + 18 chars will be the length of managed resource name). I'm not sure if we should make agent that smart just because of the conflicts. I mean, you can prevent all kinds of conflicts just by annotating a namespace and you'll explicitly know when a conflict happens so you can take action. So, I'm leaning towards keeping the name pairing between requirements and annotate them in the central cluster with UID of the namespace it exists in the remote cluster to uniquely identify them in the world of custom resources. UID of the `kube-system` could be worth adding, too, to discover in what cluster that namespace exists.

Reply:
> Let the admin of the remote cluster choose one namespace to be the default

A default that can be overridden at the namespace level seems reasonable. This would promote a pattern in which all namespaces of a remote cluster map to one namespace in the central cluster by default, but enable other patterns with administrator intervention.

It does seem like we could also support the one-to-one mapping by default if we wanted to though - perhaps this could be enabled as part of the agent's config? i.e. Whether to map every remote namespace to one central namespace if there was no annotation, or to map every remote namespace to the central namespace with the same name if there was no annotation. I would suggest that even if we choose not to support the latter case we should structure whatever flags/config file controls this such that we could add support for both if there was user demand.

> I still couldn't convince myself for this

Fair enough - I'm happy to go with identical names that will be rejected if they happen to conflict.

Reply:
> So, I'm leaning towards keeping the name pairing between requirements and annotate them in the central cluster with UID of the namespace it exists in the remote cluster to uniquely identify them in the world of custom resources. UID of the `kube-system` could be worth adding, too, to discover in what cluster that namespace exists.

A quick note - we've run into problems with this on restore since the uid of a namespace isn't durable since it's generated on create. As long as this is updated actively if the object is restored, this seems like a good approach.

Reply:
> the uid of a namespace isn't durable

Agreed. I suspect we won't be able to use the UID of `kube-system` as this would prevent us from replacing a remote cluster that was completely destroyed and recreated. I suspect assigning each agent/remote cluster a UID manually would be the better path.

Reply:
> I would suggest that even if we choose not to support the latter case we should structure whatever flags/config file controls this such that we could add support for both if there was user demand.

@negz A flag to pass `matchNamespaces: true` to enable the pairing makes sense here as an option.

Hmm restoration mechanics are interesting. The agent will have three scenarios:
* Each namespace is annotated with the target namespace.
  * No conflict for single remote cluster.
  * Could conflict if more than one cluster is connected and its namespaces have common annotations with other namespaces in other clusters.
* Each namespace targets a namespace with the same name in the central cluster.
  * No conflict for single remote cluster.
  * Could conflict if more than one cluster is connected and its namespaces have same names.
* All namespaces roll up to a default namespace.
  * Could conflict for single remote cluster if different namespaces have requirements with same name.
  * Could conflict if more than one cluster is connected to the same namespace but it's less likely to be in that situation since the default namespace selection is done explicitly during agent setup, probably by an admin.

UID of the namespace of the requirement in the remote cluster is unique enough to prevent all these conflict cases, however, it's not durable as mentioned. The name of the namespace could be considered more durable since restoration of the remote cluster will bring the namespaces back with the same name but it doesn't guarantee enough uniqueness among the N namespace * M clusters. We can actually have a unique identifier of a namespace via `name of the namespace + unique identifier of the cluster` but the problem with finding a unique identifier for the cluster is that there is none except UID of `kube-system` which doesn't meet durability standard of the restoration.

I don't love it but @negz 's suggestion of having the user manually enter a cluster identifier makes sense to achieve the uniqueness although it's not automatic as UID, i.e. user has to be careful about it. So, I'm proposing that:
* Agent will accept a cluster identifier as an installation flag like `agentIdentifier: string`
* If the flag is not provided, it will use UID of `kube-system`.

I imagine that user won't really bother supplying an identifier but in case the remote cluster was to be replaced, they can enter the UID of the `kube-system` as cluster identifier during the setup of the new remote cluster. Let's say a disaster happened and remote cluster is gone suddenly, they can look up the annotations of the requirements created by that cluster to discover what's used as cluster identifier. So we don't force them to enter an identifier, it'll have a sane default.